REVIEW: Working light client for basecoin app

ethanfrey opened the pull request.

ethanfrey commented:
Try it out.

Setup:
```
make get_vendor_deps && make build
rm -rf ~/.basecoind ~/.basecli
```

Terminal 1:
```
./build/basecoind init
# copy seed phrase into clipboard
./build/basecoind start
```

Terminal 2:
```
./build/basecli keys add demo --recover
# use password "1234567890", hardcoded in sendtx right now
# enter the seedphrase above, and 
# store the address returned as ADDR
./build/basecli account $ADDR --node=localhost:46657

TO=00CAFE00DEADBEEF00CAFE00DEADBEEF00CAFE00
./build/basecli send --to=$TO --amount=1000mycoin --name=demo --seq=0 --node=localhost:46657
./build/basecli account $TO --node=localhost:46657
./build/basecli account $ADDR --node=localhost:46657
```

Still a bit hacky, but first end-to-end test with a tendermint client.

ethanfrey commented:
Fun note... Stopping `basecoind` and starting it again leads to an error about invalid db state :(

Of course, it is also using an experimental feature branch of tendermint, so I'm not sure if it is worth writing the issue yet

ethanfrey commented:
BTW, @rigelrozanski @ebuchman can you give me any advice how to write a test that checks those above scripts work without using a bash test?

I can break this and no tests will fail. And I do not want to manually test every commit before pushing.

ethanfrey commented on x/bank/commands/sendtx.go:
```diff
@@ -0,0 +1,138 @@
+	return nil
+}
+
+func buildTx() ([]byte, error) {
```
@rigelrozanski  
this is tied to the tx struct, which can be different
for every app. How do we best keep most of the command logic in x/bank,
but allow overriding everything after buildMsg()?

I ask you, as you will be the first major user of this functionality
For now, we can just hard-code, but it is ugly.

rigelrozanski replied:
cool - I'll think about this some more - needs to constructed for applications to generically add their own structure

ethanfrey commented:
I think it is basically ready.
The whole testing is lacking a bit, but not really sure how to test the cli scripts too well, as there is little logic, just wiring inputs to calls to other libraries.

We have `tests/check_basecli.sh` to run an automated manual test (you have to interpret the results). It would be great if someone could make a proper golang version (and show me what that is supposed to look like).

Please review

schuyler commented:
I can confirm that this branch builds and runs, and I can follow @ethanfrey's instructions for testing and they work.

However, if I stop `basecoind` and restart, I get the following error:

```
$ ./basecoind start
I[02-23|20:03:28.279] Starting ABCI with Tendermint                module=main
Version > 0 but missing key header
```

Might be good to fix this before merging...?

ethanfrey commented:
This is issue #507 and has nothing to do with the cli bindings, but rather the multistore implementation.

rigelrozanski commented on client/input.go:
```diff
@@ -0,0 +1,87 @@
+	return pass, nil
+}
+
+func inputIsTty() bool {
```
could we get a few more comments around here

ethanfrey replied:
Okay, adding them

rigelrozanski commented on x/bank/commands/sendtx.go:
```diff
@@ -0,0 +1,138 @@
+			result.CheckTx.Log)
+	}
+	if result.DeliverTx.Code != uint32(0) {
+		return errors.Errorf("CheckTx failed: (%d) %s",
```
typo should be DeliverTx

ethanfrey replied:
Thanks

adrianbrink commented:
Is it correct that we are trying to merge this into `feature/sdk2-basecli`?

ebuchman commented on client/rpc/block.go:
```diff
@@ -0,0 +1,63 @@
+	if err != nil {
+		return err
+	}
+	fmt.Println(string(output))
```
the output of this is pretty gross. any idea what's going on ?

ethanfrey replied:
go-wire nightmare

ethanfrey replied:
You mean the signature as an array of ints... same as status with the pubkey?
I can make it use go-wire json.... a bit nicer but still not

ethanfrey replied:
Before:
```
./build/basecli status
{
  "node_info": {
  "pub_key": [
  18,
  24,
  175,
  75,
  128,
  241,
  196,
  70,
  192,
  62,
  146,
  211,
  201,
  156,
  153,
  122,
  69,
  169,
  116,
  12,
  206,
  247,
  243,
  151,
  8,
  229,
  60,
  131,
  144,
  226,
  127,
  157
  ],
  "listen_addr": "10.1.23.73:46656",
  "network": "test-chain-Dod558",
  "version": "0.16.0",
  "channels": "QCAhIiMwOAA=",
  "moniker": "micropower",
  "other": [
  "wire_version=0.7.2",
  "p2p_version=0.5.0",
  "consensus_version=v1/0.2.2",
  "rpc_version=0.7.0/3",
  "tx_index=on",
  "rpc_addr=tcp://0.0.0.0:46657"
  ]
  },
  "pub_key": [
  83,
  54,
  32,
  163,
  112,
  137,
  3,
  178,
  100,
  246,
  90,
  29,
  153,
  69,
  58,
  100,
  60,
  168,
  173,
  185,
  98,
  190,
  7,
  136,
  191,
  183,
  16,
  224,
  39,
  200,
  219,
  216
  ],
  "latest_block_hash": "F09AFFF7CBE65794A46528643D7A654A9FA10756",
  "latest_app_hash": "F38F97BCD4E37C09FBBBFEAF86B3770B4F00A967",
  "latest_block_height": 15,
  "latest_block_time": "2018-02-28T15:32:20.755406735+01:00",
  "syncing": false
  }
```

ethanfrey replied:
After:
```
./build/basecli status
{"node_info":{"pub_key":{"_df":"B3EF72328EAF59","_v":"EhivS4DxxEbAPpLTyZyZekWpdAzO9/OXCOU8g5Dif50="},"listen_addr":"10.1.23.73:46656","network":"test-chain-Dod558","version":"0.16.0","channels":"QCAhIiMwOAA=","moniker":"micropower","other":["wire_version=0.7.2","p2p_version=0.5.0","consensus_version=v1/0.2.2","rpc_version=0.7.0/3","tx_index=on","rpc_addr=tcp://0.0.0.0:46657"]},"pub_key":{"_df":"B3EF72328EAF59","_v":"UzYgo3CJA7Jk9lodmUU6ZDyorblivgeIv7cQ4CfI29g="},"latest_block_hash":"66760BB85DB635978C33601DF5E458816B03A8C5","latest_app_hash":"0BCDF9F6F56D31F9D0806BCFE7204EB5E3B98AF9","latest_block_height":206,"latest_block_time":"2018-02-28T15:35:38.56783532+01:00","syncing":false}
```

ebuchman commented on x/bank/commands/account.go:
```diff
@@ -0,0 +1,85 @@
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/examples/basecoin/app"
```
doesn't seem like this should be in x/bank if it needs to import basecoin!

ethanfrey replied:
there is a comment that we need to refactor handling tx type.
left as a todo for rigel to reflect what he wants since friday.
no answer yet

ebuchman commented on x/bank/commands/sendtx.go:
```diff
@@ -0,0 +1,138 @@
+	"github.com/spf13/viper"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/examples/basecoin/app" // XXX: not good
```
agreed :)

ebuchman commented:
As for tests, @zramsay can you help write the Go version of the shell tests so we can avoid shell tests? See https://github.com/cosmos/cosmos-sdk/issues/524

ethanfrey commented on client/input.go:
```diff
@@ -0,0 +1,93 @@
+	"os"
+	"strings"
+
+	"github.com/bgentry/speakeasy"
```
No idea, they seem to fulfil same purpose. Do you have a preference?
This was the package we used in cosmos-sdkV0.7 and v0.8, and no one complained...

ethanfrey commented:
Very last minute review, but i will try to address today still

ethanfrey commented on client/keys/README.md:
```diff
@@ -0,0 +1,117 @@
+# Keys CLI
+
+This is as much an example how to expose cobra/viper, as for a cli itself
+(I think this code is overkill for what go-keys needs). But please look at
```
sure thing

ethanfrey commented on client/keys/add.go:
```diff
@@ -0,0 +1,131 @@
+// Copyright © 2017 Ethan Frey
```
For now?

I removed them (guess my name doesn't need to be there), but I added this to some repos after a discussion of adding headers to all files. Maybe you can just update them to (c) All In Bits?

ethanfrey commented on client/keys.go:
```diff
@@ -0,0 +1,33 @@
+const KeyDBName = "keys"
+
+// GetKeyBase initializes a keybase based on the configuration
+func GetKeyBase(rootDir string) (keys.Keybase, error) {
```
Cleaned this up. Put the logic into client/keys/utils.go

ethanfrey commented on x/bank/commands/sendtx.go:
```diff
@@ -0,0 +1,138 @@
+	name := viper.GetString(client.FlagName)
+	info, err := keybase.Get(name)
+	if err != nil {
+		return nil, errors.WithMessage(err, "No key for name")
```
Just the "no key for name", clearer than the go-wire error

ethanfrey commented:
I think I addressed all issues.
It is far from perfect and has no proofs, but if no one else is working on this... maybe we can just roll it into develop, and let those external devs trying to use the sdk get some cli loving.

rigelrozanski commented:
got it workin with refactor! woot merging
